### PR TITLE
Feature loyal privateer

### PR DESCRIPTION
**Pull request by karlthepagan**
these are the framework changes needed to customize how privateers operate when they are full of stolen goods

### CODE REVIEW

**Comment by cyberblast** [comment]
hmm, the room additions already came in with #408 . 
Why do you want to encapsulate the worker call?

**Comment by karlthepagan** [comment]
whoops, looks like I didn't rebase #408 properly... I think i've straightened my current commits out

here's the internal feature for the worker call: https://github.com/cyberblast/screeps.ocs.internal/pull/156

**Comment by cyberblast** [comment]
I don't think this makes any sense at all. 
nextWorkAction is called ONLY when at home room... 

What do you want to achieve?

**Comment by karlthepagan** [comment]
@cyberblast I've pushed a clean reabse of this feature.

nextWorkAction is also called when the energy is full outside the home room.

The intent of my customization is to prevent privateers from repairing roads / containers in border territories and instead come home with their energy stores. See https://github.com/cyberblast/screeps.ocs.internal/pull/156

**Comment by cyberblast** [comment]
I'd propose to completely remove "repairing" instead. we've got "repair on driveby" implemented recently. You may try if that is sufficient...

**Comment by karlthepagan** [comment]
No problem. I only really issued this PR in order to justify #408. As long as the ability to customize how the privateer goes to work exists then everyone can customize it's behavior as they need.

**Comment by karlthepagan** [comment]
https://github.com/cyberblast/screeps.ocs.internal/pull/156 isn't really consequential. I just want a mechanism to extend the work behavior specifically. It's easier than rewriting the whole privateer class.

File: creep.behaviour.privateer.js [hunk 1]
```diff
@@ -16,6 +16,9 @@ module.exports = {
             logError('Creep without action/activity!\nCreep: ' + creep.name + '\ndata: ' + JSON.stringify(creep.data));
         }
     },
+    nextWorkAction: function(creep, carrySum){
```
Comment (cyberblast) [inline creep.behaviour.privateer.js]: carrySum is included in creep (as cached property creep.sum)

Comment (karlthepagan) [inline creep.behaviour.privateer.js]: Thanks, didn't see the caching.

**Comment by cyberblast** [comment]
but 

> nextWorkAction is also called when the energy is full outside the home room. 

is not correct...

**Comment by karlthepagan** [comment]
Oops, sorry I misunderstood the home room concept. Of course that's the case, but privateers operate outside of allied rooms. (robbing energy from structures, exploiting unguarded sources).

Both calling cases here are significant to me: https://github.com/karlthepagan/screeps.behaviour-action-pattern/blob/8a41d1ed52b9cc1ccd733774c164b869bbfbe4d7/creep.behaviour.privateer.js#L43

I was using privateers to great advantage in this room https://screeps.com/a/#!/room/W73S55

Unfortunately after they grabbed the energy they proceeded to repair my neighbor's roads.

**Comment by cyberblast** [comment]
No. homeRoom is always the room of its birth. The room where it will bring the energy to. its home. 
The reason why it is repairing those structures is 
https://github.com/karlthepagan/screeps.behaviour-action-pattern/blob/02f956650a706c7b1b57dedd567a4088195bb633/creep.behaviour.privateer.js#L104 

That's what I was referring to when proposed to completely remove that repairing action ;)

**Comment by cyberblast** [comment]
I'll create a PR...

**Comment by karlthepagan** [comment]
Sorry I was probably misled by some debugging statements I put in there. This makes more sense.

**Comment by cyberblast** [comment]
no worries. I'm glad to help getting things sorted

**Pull request closed without merging**